src/main/java/com/github/packageurl/PackageURL.java

```diff
@@ -433,33 +433,49 @@ private String canonicalize(boolean coordinatesOnly) {
         return purl.toString();
     }
 
-    /**
-     * Encodes the input in conformance with RFC 3986.
-     *
-     * @param input the String to encode
-     * @return an encoded String
-     */
-    private String percentEncode(final String input) {
-        return uriEncode(input, StandardCharsets.UTF_8);
-    }
+    private static String uriEncode(final String source) {
+        if (source == null || source.isEmpty()) {
+            return source;
+        }
+
+        byte[] bytes = source.getBytes(StandardCharsets.UTF_8);
+        int length = bytes.length;
+        int pos = indexOfFirstUnsafeChar(bytes);
 
-    private static String uriEncode(String source, Charset charset) {
-        if (source == null || source.length() == 0) {
+        if (pos == -1) {
             return source;
         }
 
-        StringBuilder builder = new StringBuilder();
-        for (byte b : source.getBytes(charset)) {
+        StringBuilder sb = new StringBuilder(length * 3);
+        sb.append(source, 0, pos);
+
+        for (int i = pos; i < bytes.length; i++) {
+            byte b = bytes[i];
+
             if (isUnreserved(b)) {
-                builder.append((char) b);
+                sb.append((char) b);
+            } else {
+                sb.append('%');
+                sb.append(Character.toUpperCase(Character.forDigit((b >> 4) & 0xF, 16)));
+                sb.append(Character.toUpperCase(Character.forDigit(b & 0xF, 16)));
```

Comment:
If you want to optimize this, using a common `new char[3]` and appending that buffer to the `StringBuilder` should be slightly faster, since it checks the capacity of `StringBuilder` only once.

A JMH benchmark would probably be useful.

Reply:
Performance can be increased by using `ByteBuffer::put(byte[], ...)`.

Previously, I was keeping track of the first chunk, but really you'd have to keep track of all positions and I am not sure that this extra complication is worth it.